src/localplayer.cpp

```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 				continue;
 
 			min_distance_f = distance_f;
-			new_sneak_node = p;
+			m_sneak_node = p;
+			m_sneak_node_exists = true;
 		}
 
-		bool sneak_node_found = (min_distance_f < 100000.0 * BS);
-		m_sneak_node = new_sneak_node;
-		m_sneak_node_exists = sneak_node_found;
-
-		if (sneak_node_found) {
+		if (m_sneak_node_exists) {
 			// Update saved top bounding box of sneak node
-			MapNode n = map->getNodeNoEx(m_sneak_node);
+			node = map->getNodeNoEx(m_sneak_node);
 			std::vector<aabb3f> nodeboxes;
-			n.getCollisionBoxes(nodemgr, &nodeboxes);
+			node.getCollisionBoxes(nodemgr, &nodeboxes);
 			m_sneak_node_bb_top = getTopBoundingBox(nodeboxes);
 
-			m_sneak_ladder_detected = physics_override_sneak_glitch &&
-					detectSneakLadder(map, nodemgr, floatToInt(position, BS));
-		} else {
-			m_sneak_ladder_detected = false;
+			if (physics_override_sneak_glitch) {
+				// Detect sneak ladder:
+				// Node two meters above sneak node must be solid
+				node = map->getNodeNoEx(m_sneak_node + v3s16(0,2,0),
+					&is_valid_position);
+				if (is_valid_position && nodemgr->get(node).walkable) {
+					// Node three meters above: must be non-solid
+					node = map->getNodeNoEx(m_sneak_node + v3s16(0,3,0),
+						&is_valid_position);
+					m_sneak_ladder_detected = is_valid_position &&
+						!nodemgr->get(node).walkable;
+				}
+			}
 		}
 	}
 
+	if (m_sneak_ladder_detected) {
+		// restore legacy behaviour (this makes the m_speed.Y hack necessary)
+		sneak_max = v3f(0.4 * BS, 0, 0.4 * BS);
+	}
+
+	/*
+		If sneaking, keep in range from the last walked node and don't
+		fall off from it
+	*/
+	if (could_sneak && m_sneak_node_exists) {
+		const v3f sn_f = intToFloat(m_sneak_node, BS);
+		const v3f bmin = m_sneak_node_bb_top.MinEdge;
+		const v3f bmax = m_sneak_node_bb_top.MaxEdge;
+		const v3f old_pos = position;
+
+		position.X = rangelim(position.X,
+			sn_f.X + bmin.X - sneak_max.X, sn_f.X + bmax.X + sneak_max.X);
+		position.Z = rangelim(position.Z,
+			sn_f.Z + bmin.Z - sneak_max.Z, sn_f.Z + bmax.Z + sneak_max.Z);
+
+		if (position.X != old_pos.X)
+			m_speed.X = 0;
+		if (position.Z != old_pos.Z)
+			m_speed.Z = 0;
+
+		if (m_speed.Y < 0) {
+			position.Y = MYMAX(position.Y, sn_f.Y + bmax.Y);
+			if (position.Y > old_pos.Y) {
+				CollisionInfo sn_info;
```

Comment:
> // Handle only fall damage
// (because otherwise walking against something in fast_move kills you)
[clientenvironment.cpp](https://github.com/minetest/minetest/blob/master/src/clientenvironment.cpp#L221)
Will add the collision entry for X,Z axis too.